### PR TITLE
[build] Write environment configuration file

### PR DESCRIPTION
**Pull request by imphil**
As part of Meson's configuration process initiated by `meson_init.sh` write
a file `.env` into the source directory (REPOTOP) containing environment
information as it is detected and used by Meson to build software. This
file can be used by other tools which build software to ensure that they
use the same tools as used by software built through Meson.

The `.env` file is a simple KEY=VALUE format.

Signed-off-by: Philipp Wagner <phw@lowrisc.org>

### CODE REVIEW

**Comment by imphil** [comment]
See also https://github.com/lowRISC/opentitan/pull/4165#issuecomment-731089589 for more background on this PR.

**Review by rswarbrick — APPROVED** [review]
Hmm. Looking at the Docker documentation, the `.env` files can't just be sourced by a shell (if a value has spaces, I don't think a `.env` file quotes it). If we're sticking with that format (sounds good to me), I think the commit comment needs changing, to avoid people trashing their shells :-)

Does a "consume a Docker `.env` file and generate a bash snippet to source" script exist?

**Comment by imphil** [comment]
Good point about spaces. I don't really mind about the format at all, I just remembered that Docker had a mimimal spec for those files and went with it. (I was trying to keep it in a format that would be well-defined enough to allow for parsing.)

Looking at https://gist.github.com/judy2k/7656bfe3b322d669ef75364a46327836, there are a couple reasonably easy options to get it sourced, maybe I'll leave that to the users and just remove the part of the commit message :smile:

**Comment by rswarbrick** [comment]
> Looking at https://gist.github.com/judy2k/7656bfe3b322d669ef75364a46327836, there are a couple reasonably easy options to get it sourced, maybe I'll leave that to the users and just remove the part of the commit message

My god, there are a lot of bogus "solutions" on that list. Anyway, I can see at least one IFS hack in there that should work, so I claim it's "obvious" :smile:

File: meson.build [hunk 1]
```diff
@@ -155,12 +155,17 @@ add_project_arguments(
 
 # Common program references.
 prog_python = import('python').find_installation('python3')
+prog_env = find_program('env')
+prog_git = find_program('git')
+prog_srec_cat = find_program('srec_cat')
```
Comment (lenary) [inline meson.build:160]: These two are unused?

File: meson.build [hunk 2]
```diff
@@ -203,3 +208,22 @@ top_earlgrey = declare_dependency(
...
 subdir('sw')
+
+# Write environment file
+prog_meson_write_env = meson.source_root() / 'util/meson_write_env.py'
+run_command(
+  prog_python,
+  prog_meson_write_env,
+
+  '--out-file',
+  '.env',
```
Comment (lenary) [inline meson.build:219]: Can we put this in `build-out`?

**Review by lenary — COMMENTED** [review]
Two nits, but I'm happy with the overall approach - though I'm not sure if we'd prefer to go with something we can directly `source` in common shells?

File: meson.build (hunk 2 above)
Comment (imphil) [reply meson.build]: :+1: I did go in circles with this one :-)

File: meson.build (hunk 1 above)
Comment (imphil) [reply meson.build]: srec_cat is used (in `sw/device/meson.build`), the `prog_git` is indeed unused and I just removed it.

**Comment by imphil** [comment]
> I'm not sure if we'd prefer to go with something we can directly source in common shells?

It's a bit hypothetical at the moment unfortunately.

Currently we have two likely users of this file: dvsim (Python), and the OTBN build tooling (probably a Python script, or a Makefile). Both use cases are better served with a file which is simple to parse and doesn't need to obey shell escaping/quoting semantics. I'd leave it as-is for now and revisit once we have actual users for this file. (And if you're medium-adventurous, you can simply source this file and things should be fine -- who has spaces in a toolchain path :smile:)

**Review by lenary — APPROVED** [review]

**Pull request merged**